packages/vats/src/lib-board.js

```diff
@@ -28,28 +61,114 @@ const calcCrc = (data, crcDigits) => {
+ * @see {prepareExoClassKit}
+ * @see {@link ../../SwingSet/docs/virtual-objects.md|SwingSet Virtual Objects}
+ *
+ * Hoisting this function makes defining the state type concise.
+ * @typedef {ReturnType<ReturnType<makeInit>>} BoardState
+ *
+ * @param {import('@agoric/vat-data').Baggage} baggage
+ *
+ * // TODO: use Key from @agoric/store when available
+ * @typedef {import('@endo/marshal').Passable} Key
  */
-function makeBoard(
-  initSequence = 0,
-  { prefix = DEFAULT_PREFIX, crcDigits = DEFAULT_CRC_DIGITS } = {},
-) {
+const makeInit =
+  baggage =>
+  /**
+   * @param {bigint | number} [initSequence]
+   * @param {object} [options]
+   * @param {string} [options.prefix] prefix for all ids generated
+   * @param {number} [options.crcDigits] count of digits to use in CRC at end of the id
+   */
+  (
+    initSequence = 0,
+    { prefix = DEFAULT_PREFIX, crcDigits = DEFAULT_CRC_DIGITS } = {},
+  ) => {
+    const immutable = { prefix, crcDigits };
+
+    const lastSequence = BigInt(initSequence);
+    /** @type {MapStore<string, Key>} */
+    const idToVal = provideDurableMapStore(baggage, 'idToVal');
+    /** @type {MapStore<Key, string>} */
+    const valToId = provideDurableMapStore(baggage, 'valToId');
+    const mutable = { lastSequence, idToVal, valToId };
+
+    return {
+      ...immutable,
+      ...mutable,
+    };
+  };
+
+// #region Marshaller utils
+// These utils are used by the {readonlyMarshaller, publishingMarshaller} facets
+// to serialize/unserialize. They can't return the marshallers themselves
+// because of the heap cost of maintaining those objects. Instead they make
+// transient marshallers that get GCed when the function completes.
+
+/**
+ * @param {Key} value
+ * @param {BoardState} state
+ */
+const getId = (value, state) => {
+  const { idToVal, valToId, prefix, crcDigits } = state;
+  if (!valToId.has(value)) {
+    state.lastSequence += 1n;
+    const seq = state.lastSequence;
+
+    const crcInput = `${prefix}${seq}`;
+    const crc = calcCrc(crcInput, crcDigits);
+
+    const id = `${prefix}${crc}${seq}`;
+
+    valToId.init(value, id);
+    idToVal.init(id, value);
+    return id;
+  }
+  return valToId.get(value);
+};
+
+/**
+ * @param {string} id
+ * @param {Readonly<BoardState>} state
+ */
+const getValue = (id, { prefix, crcDigits, idToVal }) => {
+  id.startsWith(prefix) || Fail`id must start with ${prefix}: ${id}`;
+  const digits = id.slice(prefix.length);
+  // the +1 is because the serial number will be at least one digit
   const DIGITS_REGEXP = new RegExp(`^[0-9]{${crcDigits + 1},}$`);
-  let lastSequence = BigInt(initSequence);
-  const idToVal = makeScalarMapStore('boardId');
-  const valToId = makeScalarMapStore('value');
+  digits.match(DIGITS_REGEXP) ||
+    Fail`id must end in at least ${q(crcDigits + 1)} digits: ${id}`;
+
+  if (idToVal.has(id)) {
+    return idToVal.get(id);
+  }
+
+  const seq = digits.slice(crcDigits);
+  const allegedCrc = digits.slice(0, crcDigits);
```

Comment:
Clarifying suggestion:
```suggestion
  const allegedCrc = digits.slice(0, crcDigits);
  const seq = digits.slice(crcDigits);
```

Reply:
no objection but what does that clarify?

Reply:
It reinforces the (prefix, CRC, unique digits) sequence by matching that ordering.